src/azure-cli-core/azure/cli/core/telemetry.py

```diff
@@ -459,6 +461,12 @@ def set_broker_info(allow_broker):
     _session.allow_broker = allow_broker
 
 
+@decorators.suppress_all_exceptions()
+def set_msal_telemetry(msal_telemetry):
+    if not _session.msal_telemetry:
+        _session.msal_telemetry = msal_telemetry
```

Comment:
If a CLI command makes several MSAL calls, only the first telemetry record is preserved.

The most special command is `az login` which calls `acquire_token_interactive` first, then several subsequent `acquire_token_silent_with_error`s:

1. `acquire_token_interactive`: initial login
2. `acquire_token_silent_with_error`: list tenants
3. `acquire_token_silent_with_error`: list subscriptions in tenant 1
4. `acquire_token_silent_with_error`: list subscriptions in tenant 2
5. ...

If step 1 fails, its telemetry will be sent and `az login` will fail. If step 1 succeeds, its telemetry will be sent, and step 2 and following steps' telemetry will be discarded, no matter if they fail or not.

For long running operation commands such as `az storage account create`, only the MSAL telemetry for the initial `PUT` request is sent. Following MSAL telemetry for `GET` polling requests is discarded.